### PR TITLE
release/0.5.0

### PR DESCRIPTION
**Pull request by Maldron**
- DEV-15926 agrian_units support
- Start adding error handling support for measurement ops.
- Support for more complex return types, failable inits, and failable fns.
- More changes for agrian_units support.
- Fix rebase issues
- Support borrowed FFI arguments for `Box`ed structs.
- Delete `ffi_consumer` crate.
- Make `ffi_common` exports a little more ergonomic.
- Fix rebase issues.
- Add error handling with `proc_macro_error`.
- cargo fmt and clean up `abort` use.
- More cleanup, early code review fixes, and so many docs.
- Fix an issue with optional types that have a custom FFI implementation.
- cargo fmt
- PR feedback.
- DEV-17287 Fix generated whitespace.
- DEV-17611 Add support for tuple structs
- Change test field name.
- ffi_core: Bump to 0.5.0
- ffi_internals: Bump to 0.5.0
- ffi_derive: Bump to 0.5.0
- ffi_common: Bump to 0.5.0
- changelog

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,34 +1,27 @@
...
-# # branch = "develop"
-
-# [patch."agrian-registry".ffi_internals]
-# path = "./ffi_internals"
-# # git = "ssh://git@github.com/agrian-inc/ffi_common.git"
-# # branch = "develop"
+[patch."agrian-registry".ffi_internals]
+path = "./ffi_internals"
+# git = "ssh://git@github.com/agrian-inc/ffi_common.git"
+# branch = "develop"
```
Comment (turboladen) [inline Cargo.toml:27]: Shouldn't these all be commented out for a release?

Comment (Maldron) [reply Cargo.toml:27]: Whoops 🤦

File: ffi_internals/src/consumer/consumer_fn.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                format!("{}.{}()", arg.name.to_string(), clone_or_borrow)
+            })
+            .collect();
+        if self.receiver != FnReceiver::None {
+            let receiver_arg = "pointer".to_string();
+            parameters.insert(0, receiver_arg);
+        }
+        parameters.join(", ")
+    }
+}
```
Comment (turboladen) [inline ffi_internals/src/consumer/consumer_fn.rs:120]: it just occurred to me, while wondering if I'd have any issues jumping in to this crate(s) to make a fix or add a feature... should we have more unit tests for these things?

File: ffi_internals/src/type_ffi.rs [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+};
+use heck::SnakeCase;
+use proc_macro2::TokenStream;
+use proc_macro_error::{abort, OptionExt};
+use quote::{format_ident, quote};
+use syn::{spanned::Spanned, Ident, Type};
+
+const STRING: &str = "String";
+const STR: &str = "str";
+const DATETIME: &str = "NaiveDateTime";
```
Comment (turboladen) [inline ffi_internals/src/type_ffi.rs:18]: Just curious... why map to `NativeDateTime`?

File: ffi_internals/src/type_ffi.rs [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+        field_name: &TokenStream,
+        has_custom_implementation: bool,
+    ) -> TokenStream {
+        // All FFIArrayT types have a `From<FFIArrayT> for Vec<T>` impl, so we can treat them all
+        // the same for the sake of native Rust assignment.
+        if self.is_vec {
+            return quote!(#field_name.into());
+        }
+
+        match self.native_type {
```
Comment (turboladen) [inline ffi_internals/src/type_ffi.rs:140]: nbd, but this might be easier to maintain (and follow?) if you pulled up the `self.is_option` logic; maybe like

```rust
match (self.native_type, self.is_option) => {
    (TypeIdentifier::Boxed(_), true) => {
    },
    (TypeIdentifier::Boxed(_), false) => {
    }
}
```

IDK

**Review by turboladen — APPROVED** [review]

File: ffi_internals/src/type_ffi.rs (hunk 3 above)
Comment (Maldron) [reply ffi_internals/src/type_ffi.rs:18]: Haven't implemented support for generics widely enough (or specifically for chrono) to use `DateTime<UTC>`, so we convert whatever kind of datetime we have to a `chrono::NaiveDateTime` (which is convertible to/from `ffi_core::datetime::TimeStamp` for FFI). It's okay as long as we convert to and expect the timestamps to be in UTC, but it's definitely not ideal. We have this in the backlog for making it better https://agrian.atlassian.net/browse/DEV-13190.

File: ffi_internals/src/consumer/consumer_fn.rs (hunk 2 above)
Comment (Maldron) [reply ffi_internals/src/consumer/consumer_fn.rs:120]: Yeah, `ffi_consumer` is horribly untested (at least without running the result against an Xcode project)... I'll add some tests to this crate for verifying the output of individual functions, and see about adding some kind of integration test that just makes sure the Swift output compiles. https://github.com/agrian-inc/ffi_common/issues/15

File: ffi_internals/src/type_ffi.rs (hunk 4 above)
Comment (Maldron) [reply ffi_internals/src/type_ffi.rs:140]: Yeah, this is...hard to follow. https://github.com/agrian-inc/ffi_common/issues/16
